File: src/reviewiq/skills.py
from __future__ import annotations

"""
ReviewIQ Skill Loader

Auto-detects languages, frameworks, and infrastructure from changed files,
then loads only the relevant skill modules into the system prompt.
This keeps token usage low — a Python/Django PR won't load React or Helm skills.

Skills are loaded from:
  1. .pr-review/skills/ (repo-level, customizable)
  2. Package templates (fallback defaults)
"""

import re
from pathlib import Path


# ── File Extension → Skill Mapping ───────────────────────────────────────────

LANGUAGE_MAP = {
    # Python
    ".py": "python", ".pyx": "python", ".pyi": "python",
    # Java / JVM
    ".java": "java", ".kt": "java", ".kts": "java", ".scala": "java", ".groovy": "java",
    # Go
    ".go": "golang",
    # TypeScript / JavaScript
    ".ts": "typescript", ".tsx": "typescript", ".js": "typescript", ".jsx": "typescript",
    ".mjs": "typescript", ".cjs": "typescript",
    # C / C++
    ".c": "cpp", ".cpp": "cpp", ".cc": "cpp", ".cxx": "cpp", ".h": "cpp",
    ".hpp": "cpp", ".hxx": "cpp",
    # Rust
    ".rs": "rust",
    # C#
    ".cs": "csharp",
    # Ruby
    ".rb": "ruby", ".erb": "ruby",
    # PHP
    ".php": "php",
    # Swift
    ".swift": "swift",
    # Kotlin (Android)
    ".kt": "kotlin",
    # SQL
    ".sql": "sql",
    # Shell
    ".sh": "shell", ".bash": "shell", ".zsh": "shell",
    # COBOL / Legacy
    ".cob": "legacy", ".cbl": "legacy", ".cpy": "legacy",
    # Fortran
    ".f": "legacy", ".f90": "legacy", ".f95": "legacy",
    # ABAP
    ".abap": "legacy",
}

FRAMEWORK_INDICATORS = {
    # Python frameworks
    "django": {"files": ["manage.py", "settings.py", "wsgi.py", "asgi.py"],
               "imports": ["django", "from django"]},
    "fastapi": {"files": [],
                "imports": ["fastapi", "from fastapi"]},
    "flask": {"files": [],
              "imports": ["flask", "from flask"]},
    "celery": {"files": ["celeryconfig.py"],
               "imports": ["celery", "from celery"]},
    "sqlalchemy": {"files": [],
                   "imports": ["sqlalchemy", "from sqlalchemy"]},
    # JS/TS frameworks
    "react": {"files": [],
              "imports": ["react", "from react", "'react'", "\"react\""]},
    "nextjs": {"files": ["next.config.js", "next.config.ts", "next.config.mjs"],
               "imports": ["next/", "from next"]},
    "express": {"files": [],
                "imports": ["express", "from express"]},
    "nestjs": {"files": [],
               "imports": ["@nestjs/"]},
    "vue": {"files": [],
            "imports": ["vue", "from vue"]},
    "angular": {"files": ["angular.json"],
                "imports": ["@angular/"]},
    # Java frameworks
    "spring": {"files": ["pom.xml", "build.gradle"],
               "imports": ["org.springframework", "spring-boot"]},
    # Ruby frameworks
    "rails": {"files": ["Gemfile", "Rakefile", "config/routes.rb"],
              "imports": ["rails", "activerecord"]},
    # .NET
    "dotnet": {"files": [],
               "imports": ["Microsoft.AspNetCore", "System.Linq", "Microsoft.EntityFrameworkCore"]},
}

DEVOPS_INDICATORS = {
    "docker": {
        "files": ["Dockerfile", "docker-compose.yml", "docker-compose.yaml",
                  ".dockerignore", "docker-compose.override.yml"],
        "extensions": [],
    },
    "kubernetes": {
        "files": [],
        "extensions": [".yaml", ".yml"],
        "content_patterns": ["apiVersion:", "kind: Deployment", "kind: Service",
                            "kind: Pod", "kind: ConfigMap", "kind: Ingress",
                            "kind: StatefulSet", "kind: DaemonSet"],
    },
    "helm": {
        "files": ["Chart.yaml", "Chart.yml", "values.yaml", "values.yml"],
        "extensions": [".tpl"],
        "dir_patterns": ["charts/", "templates/"],
    },
    "terraform": {
        "files": [],
        "extensions": [".tf", ".tfvars"],
    },
    "cicd": {
        "files": [".github/workflows/", ".gitlab-ci.yml", "Jenkinsfile",
                  ".circleci/config.yml", ".travis.yml", "azure-pipelines.yml",
                  "bitbucket-pipelines.yml"],
        "extensions": [],
    },
    "ansible": {
        "files": ["playbook.yml", "ansible.cfg", "inventory"],
        "extensions": [],
        "dir_patterns": ["roles/", "playbooks/"],
    },
}

# ── Always-loaded skills ─────────────────────────────────────────────────────

ALWAYS_LOAD = ["commandments", "security", "scalability", "stability", "maintainability", "performance"]

FINTECH_INDICATORS = {
    "imports": [
        "stripe", "razorpay", "paypal", "braintree", "adyen", "square",
        "plaid", "dwolla", "paytm", "phonepe", "cashfree", "payu",
        "payment", "checkout", "billing", "invoice", "subscription",
        "ledger", "accounting", "journal_entry", "double_entry",
        "loan", "emi", "amortization", "disbursement", "repayment",
        "interest_rate", "apr", "prepayment", "moratorium",
        "insurance", "policy", "premium", "claim", "underwriting",
        "coverage", "endorsement", "actuary",
        "kyc", "aml", "pci", "compliance", "sanctions",
        "bank_account", "ifsc", "iban", "swift", "ach", "nach",
        "credit_score", "bureau", "cibil", "experian", "equifax",
        "wallet", "upi", "neft", "rtgs", "imps",
    ],
    "files": [
        "payment", "checkout", "billing", "invoice", "subscription",
        "loan", "emi", "lending", "disburs", "repay", "collect",
        "insurance", "policy", "claim", "premium", "underwrit",
        "ledger", "journal", "accounting", "reconcil",
        "kyc", "aml", "compliance", "sanction",
        "wallet", "transfer", "payout", "refund", "settlement",
    ],
}


# ── Detection Engine ─────────────────────────────────────────────────────────

def detect_skills(changed_files: list[str], file_contents: str = "") -> dict:
    """
    Detect which skills are relevant based on changed files and their contents.

    Returns:
        {
            "languages": ["python", "typescript"],
            "frameworks": ["django", "react"],
            "devops": ["docker", "kubernetes", "helm"],
            "always": ["commandments", "security", "scalability", "stability"],
        }
    """
    result = {
        "languages": set(),
        "frameworks": set(),
        "devops": set(),
        "fintech": False,
        "always": ALWAYS_LOAD,
    }

    content_lower = file_contents.lower()

    for filepath in changed_files:
        path = Path(filepath)
        ext = path.suffix.lower()
        name = path.name.lower()

        # Language detection
        if ext in LANGUAGE_MAP:
            result["languages"].add(LANGUAGE_MAP[ext])

        # DevOps detection by filename
        for devops_type, indicators in DEVOPS_INDICATORS.items():
            for pattern in indicators.get("files", []):
                if name == pattern.lower() or filepath.lower().endswith(pattern.lower()):
                    result["devops"].add(devops_type)
            for dev_ext in indicators.get("extensions", []):
                if ext == dev_ext:
                    result["devops"].add(devops_type)
            for dir_pat in indicators.get("dir_patterns", []):
                if dir_pat.lower() in filepath.lower():
                    result["devops"].add(devops_type)

        # Framework detection by filename
        for framework, indicators in FRAMEWORK_INDICATORS.items():
            for pattern in indicators.get("files", []):
                if name == pattern.lower() or filepath.lower().endswith(pattern.lower()):
                    result["frameworks"].add(framework)

    # Content-based detection (imports, patterns)
    if file_contents:
        # Framework detection by imports
        for framework, indicators in FRAMEWORK_INDICATORS.items():
            for imp in indicators.get("imports", []):
                if imp.lower() in content_lower:
                    result["frameworks"].add(framework)

        # Kubernetes content patterns
        k8s_indicators = DEVOPS_INDICATORS.get("kubernetes", {})
        for pattern in k8s_indicators.get("content_patterns", []):
            if pattern.lower() in content_lower:
                result["devops"].add("kubernetes")

    # Fintech detection (content-based)
    if file_contents:
        for pattern in FINTECH_INDICATORS["imports"]:
            if pattern.lower() in content_lower:
                result["fintech"] = True
                break
    if not result["fintech"]:
        for pattern in FINTECH_INDICATORS["files"]:
            for filepath in changed_files:
                if pattern.lower() in filepath.lower():
                    result["fintech"] = True
                    break
            if result["fintech"]:
                break

    # Convert sets to sorted lists for deterministic ordering
    result["languages"] = sorted(result["languages"])
    result["frameworks"] = sorted(result["frameworks"])
    result["devops"] = sorted(result["devops"])

    return result


# ── Skill File Loading ───────────────────────────────────────────────────────

def _skill_dirs() -> list[Path]:
    """Return skill directories in priority order (repo first, package fallback)."""
    dirs = []
    repo_skills = Path(".pr-review/skills")
    if repo_skills.exists():
        dirs.append(repo_skills)
    pkg_skills = Path(__file__).parent / "templates" / "skills"
    if pkg_skills.exists():
        dirs.append(pkg_skills)
    return dirs


def _load_skill_file(name: str) -> str | None:
    """Load a skill file by name from the first available directory."""
    for skill_dir in _skill_dirs():
        path = skill_dir / f"{name}.md"
        if path.exists():
            return path.read_text()
    return None


def load_skills(detected: dict) -> str:
    """
    Load all relevant skill files and compose them into a single prompt section.

    Only loads what's needed — a Python/Django PR won't load React or Helm skills.
    """
    sections = []

    # Always-loaded skills
    for skill_name in detected["always"]:
        content = _load_skill_file(skill_name)
        if content:
            sections.append(content)

    # Language skills
    if detected["languages"]:
        content = _load_skill_file("languages")
        if content:
            # Extract only relevant language sections
            relevant = _extract_sections(content, detected["languages"])
            if relevant:
                sections.append(f"# Language-Specific Review Rules\n\n{relevant}")

    # Framework skills
    if detected["frameworks"]:
        content = _load_skill_file("frameworks")
        if content:
            relevant = _extract_sections(content, detected["frameworks"])
            if relevant:
                sections.append(f"# Framework-Specific Review Rules\n\n{relevant}")

    # DevOps skills
    if detected["devops"]:
        content = _load_skill_file("devops")
        if content:
            relevant = _extract_sections(content, detected["devops"])
            if relevant:
                sections.append(f"# DevOps Review Rules\n\n{relevant}")

    # Fintech skills
    if detected.get("fintech"):
        content = _load_skill_file("fintech")
        if content:
            sections.append(content)

    if not sections:
        return ""

    header = "---\n\n# REVIEW SKILLS (auto-loaded based on detected files)\n\n"
    skills_loaded = []
    skills_loaded.extend(detected["always"])
    skills_loaded.extend(detected["languages"])
    skills_loaded.extend(detected["frameworks"])
    skills_loaded.extend(detected["devops"])
    if detected.get("fintech"):
        skills_loaded.append("fintech")

    header += f"**Skills loaded**: {', '.join(skills_loaded)}\n\n"
    return header + "\n\n---\n\n".join(sections)


def _extract_sections(content: str, keys: list[str]) -> str:
    """
    Extract only the sections matching the given keys from a skill file.

    Sections are delimited by ## headers. A section matches if any key
    appears in the header (case-insensitive).
    """
    sections = re.split(r"(?=^## )", content, flags=re.MULTILINE)
    matched = []

    for section in sections:
        header_match = re.match(r"^## (.+)$", section, re.MULTILINE)
        if not header_match:
            continue
        header = header_match.group(1).lower()
        for key in keys:
            if key.lower() in header:
                matched.append(section.strip())
                break

    return "\n\n".join(matched)

File: internal/skills/skills.go
// Package skills auto-detects languages, frameworks, and domains from changed files
// and loads only the relevant skill modules into the system prompt.
package skills

import (
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// ── Mappings ────────────────────────────────────────────────────────────────

var languageMap = map[string]string{
	".py": "python", ".pyx": "python", ".pyi": "python",
	".java": "java", ".kt": "java", ".kts": "java", ".scala": "java", ".groovy": "java",
	".go": "golang",
	".ts": "typescript", ".tsx": "typescript", ".js": "typescript", ".jsx": "typescript",
	".mjs": "typescript", ".cjs": "typescript",
	".c": "cpp", ".cpp": "cpp", ".cc": "cpp", ".cxx": "cpp", ".h": "cpp",
	".hpp": "cpp", ".hxx": "cpp",
	".rs": "rust", ".cs": "csharp",
	".rb": "ruby", ".erb": "ruby",
	".php": "php", ".swift": "swift",
	".sql": "sql",
	".sh": "shell", ".bash": "shell", ".zsh": "shell",
	".cob": "legacy", ".cbl": "legacy", ".cpy": "legacy",
	".f": "legacy", ".f90": "legacy", ".f95": "legacy",
	".abap": "legacy",
}

type indicator struct {
	imports []string
	files   []string
}

var frameworkIndicators = map[string]indicator{
	"django":  {imports: []string{"django", "from django"}, files: []string{"manage.py", "settings.py", "wsgi.py", "asgi.py"}},
	"fastapi": {imports: []string{"fastapi", "from fastapi"}},
	"flask":   {imports: []string{"flask", "from flask"}},
	"react":   {imports: []string{"react", "from react", "'react'", "\"react\""}},
	"nextjs":  {imports: []string{"next/", "from next"}, files: []string{"next.config.js", "next.config.ts", "next.config.mjs"}},
	"express": {imports: []string{"express", "from express"}},
	"nestjs":  {imports: []string{"@nestjs/"}},
	"vue":     {imports: []string{"vue", "from vue"}},
	"angular": {imports: []string{"@angular/"}, files: []string{"angular.json"}},
	"spring":  {imports: []string{"org.springframework", "spring-boot"}, files: []string{"pom.xml", "build.gradle"}},
	"rails":   {imports: []string{"rails", "activerecord"}, files: []string{"Gemfile", "Rakefile"}},
	"dotnet":  {imports: []string{"Microsoft.AspNetCore", "System.Linq", "Microsoft.EntityFrameworkCore"}},
}

type devopsIndicator struct {
	files           []string
	extensions      []string
	contentPatterns []string
	dirPatterns     []string
}

var devopsIndicators = map[string]devopsIndicator{
	"docker":     {files: []string{"Dockerfile", "docker-compose.yml", "docker-compose.yaml", ".dockerignore"}},
	"kubernetes": {extensions: []string{".yaml", ".yml"}, contentPatterns: []string{"apiVersion:", "kind: Deployment", "kind: Service", "kind: Pod", "kind: ConfigMap", "kind: Ingress"}},
	"helm":       {files: []string{"Chart.yaml", "Chart.yml", "values.yaml", "values.yml"}, extensions: []string{".tpl"}, dirPatterns: []string{"charts/", "templates/"}},
	"terraform":  {extensions: []string{".tf", ".tfvars"}},
	"cicd":       {files: []string{".github/workflows/", ".gitlab-ci.yml", "Jenkinsfile", ".circleci/config.yml", ".travis.yml", "azure-pipelines.yml"}},
	"ansible":    {files: []string{"playbook.yml", "ansible.cfg", "inventory"}, dirPatterns: []string{"roles/", "playbooks/"}},
}

type domainDef struct {
	imports  []string
	files    []string
	skillFile string
}

var domainIndicators = map[string]domainDef{
	"fintech": {
		imports:  []string{"stripe", "razorpay", "paypal", "braintree", "adyen", "payment", "checkout", "billing", "invoice", "ledger", "accounting", "loan", "emi", "amortization", "disbursement", "insurance", "policy", "premium", "claim", "underwriting", "kyc", "aml", "pci", "wallet", "upi"},
		files:    []string{"payment", "checkout", "billing", "loan", "emi", "lending", "insurance", "policy", "claim", "ledger", "reconcil", "kyc", "aml", "wallet", "refund", "settlement"},
		skillFile: "fintech",
	},
	"india_regulatory": {
		imports:  []string{"upi", "npci", "nach", "ecs", "neft", "rtgs", "imps", "aadhaar", "uidai", "digilocker", "ckyc", "ekyc", "ifsc", "rbi", "nbfc", "cersai", "razorpay", "paytm", "phonepe", "cashfree", "account_aggregator"},
		files:    []string{"upi", "nach", "ecs", "neft", "rtgs", "imps", "aadhaar", "ekyc", "ckyc", "nbfc", "rbi", "mandate"},
		skillFile: "india-regulatory",
	},
	"credit_bureau": {
		imports:  []string{"cibil", "transunion", "experian", "equifax", "crif", "credit_score", "credit_report", "bureau", "credit_bureau"},
		files:    []string{"bureau", "cibil", "credit_score", "credit_report"},
		skillFile: "credit-bureau",
	},
	"fraud": {
		imports:  []string{"fraud", "risk_score", "risk_engine", "velocity_check", "device_fingerprint", "anti_fraud", "aml", "money_laundering", "chargeback", "dispute"},
		files:    []string{"fraud", "risk_engine", "risk_score", "anti_fraud", "velocity", "chargeback", "dispute"},
		skillFile: "fraud",
	},
	"notifications": {
		imports:  []string{"sms", "twilio", "msg91", "kaleyra", "gupshup", "sendgrid", "ses", "mailgun", "fcm", "apns", "push_notification", "whatsapp", "dlt", "trai"},
		files:    []string{"notification", "sms", "email_service", "push", "whatsapp", "communication"},
		skillFile: "notifications",
	},
	"financial_microservices": {
		imports:  []string{"saga", "compensat", "outbox", "event_sourcing", "distributed_transaction", "circuit_breaker", "kafka", "rabbitmq", "ledger_service", "payment_service"},
		files:    []string{"saga", "orchestrat", "compensat", "outbox", "event_store", "ledger_service", "settlement"},
		skillFile: "financial-microservices",
	},
	"data_privacy": {
		imports:  []string{"gdpr", "ccpa", "dpdp", "privacy", "consent", "data_subject", "erasure", "anonymiz", "pseudonymiz", "pii", "personal_data", "data_protection"},
		files:    []string{"privacy", "consent", "gdpr", "ccpa", "dpdp", "anonymiz", "pii", "data_protection", "data_deletion"},
		skillFile: "data-privacy",
	},
	"airflow": {
		imports:  []string{"airflow", "from airflow", "DAG", "PythonOperator", "BashOperator", "TaskGroup", "XCom", "BaseOperator", "dag_id", "schedule_interval"},
		files:    []string{"dag", "airflow", "dags/", "pipeline"},
		skillFile: "airflow",
	},
	"kafka": {
		imports:  []string{"kafka", "confluent_kafka", "kafka-python", "aiokafka", "KafkaProducer", "KafkaConsumer", "kafka.clients", "org.apache.kafka", "kafkajs", "node-rdkafka"},
		files:    []string{"kafka", "producer", "consumer", "topic", "broker"},
		skillFile: "kafka",
	},
	"messaging": {
		imports:  []string{"rabbitmq", "amqp", "pika", "celery", "kombu", "bullmq", "bull", "amqplib", "sqs", "SendMessage", "ReceiveMessage", "rq", "redis_queue", "huey"},
		files:    []string{"rabbitmq", "celery", "worker", "queue", "tasks.py", "celeryconfig", "sqs", "bullmq"},
		skillFile: "messaging",
	},
	"sql": {
		imports:  []string{"select ", "insert into", "update ", "delete from", " join ", "cursor.execute", "db.query", "conn.exec", "executemany"},
		files:    []string{".sql"},
		skillFile: "sql",
	},
	"migrations": {
		imports:  []string{"alembic", "from alembic", "flyway", "liquibase", "knex.migrate", "prisma migrate", "ActiveRecord::Migration"},
		files:    []string{"migrations/", "db/migrate/", "alembic/versions/", "flyway/", "liquibase/", "_migration.py", ".changeset.xml", "schema_migrations"},
		skillFile: "migrations",
	},
	"orm": {
		imports:  []string{"django.db", "sqlalchemy", "from sqlalchemy", "prisma", "@prisma/client", "typeorm", "sequelize", "mongoose", "gorm", "hibernate", "@entity", "javax.persistence", "jakarta.persistence", "spring-data"},
		files:    []string{"models.py", "entity.ts", "entity.js", ".repository.ts", ".repository.java"},
		skillFile: "orm",
	},
	"transactions": {
		imports:  []string{"transaction.atomic", "@transactional", "begintx", "select for update", "for no key update", "pg_advisory_lock", "advisory_xact_lock", "savepoint ", "db.transaction(", "session.starttransaction"},
		skillFile: "transactions",
	},
	"postgres": {
		imports:  []string{"psycopg2", "asyncpg", "node-postgres", "from pg ", "pgx", "gorm.io/driver/postgres", "jsonb", "on conflict", "concurrently", "tsvector", "pg_trgm", "pg_stat_statements"},
		files:    []string{"postgres", "postgresql"},
		skillFile: "postgres",
	},
	"redis": {
		imports:  []string{"redis", "ioredis", "go-redis", "lettuce", "jedis", "stackexchange.redis", "setnx", "xadd", "xreadgroup", "subscribe", "redis.lua", "redis.eval"},
		files:    []string{"redis"},
		skillFile: "redis",
	},
	"mongodb": {
		imports:  []string{"pymongo", "mongoose", "mongodb", "mongoclient", "@document", "$match", "$lookup", "$group", "$unwind", "findoneandupdate", "aggregate(", "bulkwrite"},
		files:    []string{"mongo"},
		skillFile: "mongodb",
	},
	"elasticsearch": {
		imports:  []string{"elasticsearch", "@elastic/elasticsearch", "opensearch", "org.elasticsearch.client", "_search", "_bulk", "_mapping", "indices.put_mapping", "search_after"},
		files:    []string{"elastic", "opensearch"},
		skillFile: "elasticsearch",
	},
}

var alwaysLoad = []string{"commandments", "security", "scalability", "stability", "maintainability", "performance"}

// ── Detection ───────────────────────────────────────────────────────────────

type Detected struct {
	Languages  []string
	Frameworks []string
	DevOps     []string
	Domains    []string
	Always     []string
}

func Detect(changedFiles []string, fileContents string) Detected {
	langs := map[string]bool{}
	fws := map[string]bool{}
	devops := map[string]bool{}
	domains := map[string]bool{}
	contentLower := strings.ToLower(fileContents)

	for _, fp := range changedFiles {
		ext := strings.ToLower(filepath.Ext(fp))
		name := strings.ToLower(filepath.Base(fp))
		fpLower := strings.ToLower(fp)

		if lang, ok := languageMap[ext]; ok {
			langs[lang] = true
		}
		for dtype, ind := range devopsIndicators {
			for _, p := range ind.files {
				if name == strings.ToLower(p) || strings.HasSuffix(fpLower, strings.ToLower(p)) {
					devops[dtype] = true
				}
			}
			for _, e := range ind.extensions {
				if ext == e {
					devops[dtype] = true
				}
			}
			for _, d := range ind.dirPatterns {
				if strings.Contains(fpLower, strings.ToLower(d)) {
					devops[dtype] = true
				}
			}
		}
		for fw, ind := range frameworkIndicators {
			for _, p := range ind.files {
				if name == strings.ToLower(p) || strings.HasSuffix(fpLower, strings.ToLower(p)) {
					fws[fw] = true
				}
			}
		}
	}

	if fileContents != "" {
		for fw, ind := range frameworkIndicators {
			for _, imp := range ind.imports {
				if strings.Contains(contentLower, strings.ToLower(imp)) {
					fws[fw] = true
					break
				}
			}
		}
		for _, di := range devopsIndicators {
			for _, p := range di.contentPatterns {
				if strings.Contains(contentLower, strings.ToLower(p)) {
					devops["kubernetes"] = true
					break
				}
			}
		}
		for dname, dd := range domainIndicators {
			found := false
			for _, imp := range dd.imports {
				if strings.Contains(contentLower, strings.ToLower(imp)) {
					found = true
					break
				}
			}
			if !found {
				for _, fp := range dd.files {
					for _, cf := range changedFiles {
						if strings.Contains(strings.ToLower(cf), strings.ToLower(fp)) {
							found = true
							break
						}
					}
					if found {
						break
					}
				}
			}
			if found {
				domains[dname] = true
			}
		}
	}

	return Detected{
		Languages:  sortedKeys(langs),
		Frameworks: sortedKeys(fws),
		DevOps:     sortedKeys(devops),
		Domains:    sortedKeys(domains),
		Always:     alwaysLoad,
	}
}

// ── Loading ─────────────────────────────────────────────────────────────────

func skillDirs() []string {
	var dirs []string
	if info, err := os.Stat(filepath.Join(".pr-review", "skills")); err == nil && info.IsDir() {
		dirs = append(dirs, filepath.Join(".pr-review", "skills"))
	}
	return dirs
}

func loadSkillFile(name string) string {
	for _, dir := range skillDirs() {
		data, err := os.ReadFile(filepath.Join(dir, name+".md"))
		if err == nil {
			return string(data)
		}
	}
	return ""
}

func LoadSkills(d Detected) string {
	var sections []string

	for _, name := range d.Always {
		if c := loadSkillFile(name); c != "" {
			sections = append(sections, c)
		}
	}
	if len(d.Languages) > 0 {
		if c := loadSkillFile("languages"); c != "" {
			if rel := extractSections(c, d.Languages); rel != "" {
				sections = append(sections, "# Language-Specific Review Rules\n\n"+rel)
			}
		}
	}
	if len(d.Frameworks) > 0 {
		if c := loadSkillFile("frameworks"); c != "" {
			if rel := extractSections(c, d.Frameworks); rel != "" {
				sections = append(sections, "# Framework-Specific Review Rules\n\n"+rel)
			}
		}
	}
	if len(d.DevOps) > 0 {
		if c := loadSkillFile("devops"); c != "" {
			if rel := extractSections(c, d.DevOps); rel != "" {
				sections = append(sections, "# DevOps Review Rules\n\n"+rel)
			}
		}
	}
	for _, dname := range d.Domains {
		if dd, ok := domainIndicators[dname]; ok {
			if c := loadSkillFile(dd.skillFile); c != "" {
				sections = append(sections, c)
			}
		}
	}

	if len(sections) == 0 {
		return ""
	}

	var loaded []string
	loaded = append(loaded, d.Always...)
	loaded = append(loaded, d.Languages...)
	loaded = append(loaded, d.Frameworks...)
	loaded = append(loaded, d.DevOps...)
	for _, dname := range d.Domains {
		if dd, ok := domainIndicators[dname]; ok {
			loaded = append(loaded, dd.skillFile)
		}
	}

	header := "---\n\n# REVIEW SKILLS (auto-loaded based on detected files)\n\n"
	header += "**Skills loaded**: " + strings.Join(loaded, ", ") + "\n\n"
	return header + strings.Join(sections, "\n\n---\n\n")
}

// ── Helpers ─────────────────────────────────────────────────────────────────

var sectionSplitter = regexp.MustCompile(`(?m)^## `)

func extractSections(content string, keys []string) string {
	headerRe := regexp.MustCompile(`(?m)^## (.+)$`)
	allMatches := headerRe.FindAllStringIndex(content, -1)

	var matched []string
	for i, idx := range allMatches {
		end := len(content)
		if i+1 < len(allMatches) {
			end = allMatches[i+1][0]
		}
		section := content[idx[0]:end]
		headerMatch := headerRe.FindStringSubmatch(section)
		if len(headerMatch) < 2 {
			continue
		}
		headerLower := strings.ToLower(headerMatch[1])
		for _, key := range keys {
			if strings.Contains(headerLower, strings.ToLower(key)) {
				matched = append(matched, strings.TrimSpace(section))
				break
			}
		}
	}
	return strings.Join(matched, "\n\n")
}

func sortedKeys(m map[string]bool) []string {
	var out []string
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}
